qiskit/circuit/library/evolution/hamiltonian_gate.py

```diff
@@ -0,0 +1,136 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Gate described by the time evolution of a Hermitian Hamiltonian operator.
+"""
+
+from numbers import Number
+import numpy
+
+from qiskit.circuit import Gate, QuantumCircuit, ParameterExpression
+from qiskit.quantum_info.operators.predicates import matrix_equal
+from qiskit.quantum_info.operators.predicates import is_hermitian_matrix
+from qiskit.extensions.unitary import UnitaryGate
+from qiskit.circuit.exceptions import CircuitError
+
+
+class HamiltonianGate(Gate):
+    r"""Class for representing evolution by a Hermitian Hamiltonian operator as a gate.
+
+    This gate resolves to a :class:`.UnitaryGate` with definition :math:`U(t) = \exp(-i t  H)`,
+    which can be decomposed into basis gates if it is 2 qubits or less, or simulated directly for
+    more qubits.
+    """
+
+    def __init__(self, data, time, label=None):
+        """
+        Args:
+            data (matrix or Operator): a hermitian operator.
+            time (float): time evolution parameter.
+            label (str): unitary name for backend [Default: None].
+
+        Raises:
+            ValueError: if input data is not an N-qubit unitary operator.
+        """
+        if hasattr(data, "to_matrix"):
+            # If input is Gate subclass or some other class object that has
+            # a to_matrix method this will call that method.
+            data = data.to_matrix()
+        elif hasattr(data, "to_operator"):
+            # If input is a BaseOperator subclass this attempts to convert
+            # the object to an Operator so that we can extract the underlying
+            # numpy matrix from `Operator.data`.
+            data = data.to_operator().data
```

Comment:
The `hamiltonian_gate.py` was just moved from `qiskit/extensions` to here, I didn't actually do any modifications to the code 🙂 I think the diff might not have picked it up correctly since the old file still exists to deprecate the old import location... so the right diff would be: moved `hamiltonian_gate.py` and created a new file in the same old location to make it importable. Do you know if there's a way to make git follow this? Maybe explicitly do two commits, one for the move one for the file creation?

If there's no way to do this, then we could also add your suggestions here, otherwise it might be better to keep the file as is 🙂

Reply:
I included all the suggested changes on a different branch for a separate PR 🙂